File: modules/rain-alerts/ios/RainAlertModule.swift
import Foundation
import ExpoModulesCore
import BackgroundTasks
import UserNotifications
import CoreLocation

public class RainAlertModule: Module {
  private let locationManager = CLLocationManager()
  private let userDefaults = UserDefaults.standard
  private let cooldownKey = "rainAlertLastAlertTimestamp"
  private let tileIdKey = "rainAlertLastTileId"
  
  public func definition() -> ModuleDefinition {
    Name("RainAlert")
    
    Function("initialize") { (leadMinutes: Int, intensityThreshold: String) in
      self.scheduleBackgroundRefresh()
      return true
    }
    
    Function("handlePushPayload") { (payload: [String: Any]) -> Bool in
      return self.processPushPayload(payload)
    }
    
    Function("canShowAlert") { () -> Bool in
      return self.checkCooldown()
    }
    
    Function("recordAlert") { () in
      self.recordAlertTimestamp()
    }
    
    Function("getRemainingCooldown") { () -> Int in
      return self.getRemainingCooldownSeconds()
    }
    
    Function("requestPermissions") { () async -> Bool in
      return await self.requestNotificationAndLocationPermissions()
    }
    
    Function("scheduleBackgroundTasks") { () in
      self.scheduleBackgroundRefresh()
    }
  }
  
  private func scheduleBackgroundRefresh() {
    let request = BGAppRefreshTaskRequest(identifier: "com.subhanify.rainAlertRefresh")
    request.earliestBeginDate = Date(timeIntervalSinceNow: 3600)
    
    do {
      try BGTaskScheduler.shared.submit(request)
      print("Rain alert background refresh scheduled")
    } catch {
      print("Failed to schedule background refresh: \(error)")
    }
  }
  
  private func processPushPayload(_ payload: [String: Any]) -> Bool {
    guard payload["type"] as? String == "RAIN_START" else {
      return false
    }
    
    if !checkCooldown() {
      print("Rain alert suppressed due to cooldown")
      return false
    }
    
    let leadMinutes = payload["leadMinutes"] as? Int ?? 0
    let intensity = payload["intensity"] as? String ?? "light"
    let phrase = payload["phrase"] as? String ?? "Rain has started nearby"
    let dua = payload["dua"] as? String ?? "اللَّهُمَّ صَيِّبًا نَافِعًا"
    
    showNotification(
      title: phrase,
      body: "\(dua) — O Allah, a beneficial downpour.",
      userInfo: payload
    )
    
    recordAlertTimestamp()
    return true
  }
  
  private func checkCooldown() -> Bool {
    guard let lastAlert = userDefaults.object(forKey: cooldownKey) as? Date else {
      return true
    }
    
    let elapsed = Date().timeIntervalSince(lastAlert)
    return elapsed >= 3600
  }
  
  private func getRemainingCooldownSeconds() -> Int {
    guard let lastAlert = userDefaults.object(forKey: cooldownKey) as? Date else {
      return 0
    }
    
    let elapsed = Date().timeIntervalSince(lastAlert)
    let remaining = max(0, 3600 - elapsed)
    return Int(ceil(remaining))
  }
  
  private func recordAlertTimestamp() {
    userDefaults.set(Date(), forKey: cooldownKey)
    userDefaults.synchronize()
  }
  
  private func showNotification(title: String, body: String, userInfo: [String: Any]) {
    let content = UNMutableNotificationContent()
    content.title = title
    content.body = body
    content.sound = .default
    content.categoryIdentifier = "RAIN_ALERT"
    content.userInfo = userInfo
    
    let request = UNNotificationRequest(
      identifier: UUID().uuidString,
      content: content,
      trigger: nil
    )
    
    UNUserNotificationCenter.current().add(request) { error in
      if let error = error {
        print("Failed to show rain alert notification: \(error)")
      } else {
        print("Rain alert notification shown")
      }
    }
  }
  
  private func requestNotificationAndLocationPermissions() async -> Bool {
    let notificationCenter = UNUserNotificationCenter.current()
    
    do {
      let granted = try await notificationCenter.requestAuthorization(options: [.alert, .sound, .badge])
      
      if !granted {
        print("Notification permission denied")
        return false
      }
      
      let locationGranted = await requestLocationPermission()
      return locationGranted
      
    } catch {
      print("Failed to request permissions: \(error)")
      return false
    }
  }
  
  private func requestLocationPermission() async -> Bool {
    return await withCheckedContinuation { continuation in
      locationManager.requestWhenInUseAuthorization()
      
      DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
        let status = self.locationManager.authorizationStatus
        continuation.resume(returning: status == .authorizedWhenInUse || status == .authorizedAlways)
      }
    }
  }
}

File: modules/rain-alerts/android/src/main/java/com/subhanify/rainalert/RainAlertModule.kt
package com.subhanify.rainalert

import android.content.Context
import android.content.SharedPreferences
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.work.*
import com.google.firebase.messaging.RemoteMessage
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.TimeUnit

class RainAlertModule : Module() {
  private lateinit var preferences: SharedPreferences
  
  override fun definition() = ModuleDefinition {
    Name("RainAlert")
    
    onCreate {
      preferences = appContext.reactContext?.getSharedPreferences(
        "rain_alert_prefs",
        Context.MODE_PRIVATE
      ) ?: throw IllegalStateException("Context not available")
    }
    
    Function("initialize") { leadMinutes: Int, intensityThreshold: String ->
      schedulePeriodicWork()
      true
    }
    
    Function("handlePushPayload") { payload: Map<String, Any> ->
      processPushPayload(payload)
    }
    
    Function("canShowAlert") {
      checkCooldown()
    }
    
    Function("recordAlert") {
      recordAlertTimestamp()
    }
    
    Function("getRemainingCooldown") {
      getRemainingCooldownSeconds()
    }
    
    Function("requestPermissions") {
      true
    }
    
    Function("schedulePeriodicWork") {
      schedulePeriodicWork()
    }
    
    Function("cancelPeriodicWork") {
      cancelPeriodicWork()
    }
  }
  
  private fun schedulePeriodicWork() {
    val context = appContext.reactContext ?: return
    
    val constraints = Constraints.Builder()
      .setRequiredNetworkType(NetworkType.CONNECTED)
      .build()
    
    val workRequest = PeriodicWorkRequestBuilder<RainAlertWorker>(
      15, TimeUnit.MINUTES,
      5, TimeUnit.MINUTES
    )
      .setConstraints(constraints)
      .setBackoffCriteria(
        BackoffPolicy.EXPONENTIAL,
        WorkRequest.MIN_BACKOFF_MILLIS,
        TimeUnit.MILLISECONDS
      )
      .build()
    
    WorkManager.getInstance(context).enqueueUniquePeriodicWork(
      "rain_alert_check",
      ExistingPeriodicWorkPolicy.KEEP,
      workRequest
    )
  }
  
  private fun cancelPeriodicWork() {
    val context = appContext.reactContext ?: return
    WorkManager.getInstance(context).cancelUniqueWork("rain_alert_check")
  }
  
  private fun processPushPayload(payload: Map<String, Any>): Boolean {
    if (payload["type"] != "RAIN_START") {
      return false
    }
    
    if (!checkCooldown()) {
      android.util.Log.d("RainAlert", "Alert suppressed due to cooldown")
      return false
    }
    
    val leadMinutes = payload["leadMinutes"] as? Int ?: 0
    val intensity = payload["intensity"] as? String ?: "light"
    val phrase = payload["phrase"] as? String ?: "Rain has started nearby"
    val dua = payload["dua"] as? String ?: "اللَّهُمَّ صَيِّبًا نَافِعًا"
    
    showNotification(
      title = phrase,
      body = "$dua — O Allah, a beneficial downpour.",
      payload = payload
    )
    
    recordAlertTimestamp()
    return true
  }
  
  private fun checkCooldown(): Boolean {
    val lastAlert = preferences.getLong("last_alert_timestamp", 0)
    if (lastAlert == 0L) {
      return true
    }
    
    val elapsed = (System.currentTimeMillis() - lastAlert) / 1000
    return elapsed >= 3600
  }
  
  private fun getRemainingCooldownSeconds(): Int {
    val lastAlert = preferences.getLong("last_alert_timestamp", 0)
    if (lastAlert == 0L) {
      return 0
    }
    
    val elapsed = (System.currentTimeMillis() - lastAlert) / 1000
    val remaining = maxOf(0, 3600 - elapsed)
    return remaining.toInt()
  }
  
  private fun recordAlertTimestamp() {
    preferences.edit()
      .putLong("last_alert_timestamp", System.currentTimeMillis())
      .apply()
  }
  
  private fun showNotification(title: String, body: String, payload: Map<String, Any>) {
    val context = appContext.reactContext ?: return
    
    val notification = NotificationCompat.Builder(context, "RAIN_ALERT")
      .setContentTitle(title)
      .setContentText(body)
      .setSmallIcon(android.R.drawable.ic_dialog_info)
      .setPriority(NotificationCompat.PRIORITY_HIGH)
      .setCategory(NotificationCompat.CATEGORY_REMINDER)
      .setAutoCancel(true)
      .addAction(
        android.R.drawable.ic_menu_view,
        "Read Dua",
        null
      )
      .addAction(
        android.R.drawable.ic_lock_silent_mode_off,
        "Play Audio",
        null
      )
      .build()
    
    try {
      NotificationManagerCompat.from(context).notify(
        System.currentTimeMillis().toInt(),
        notification
      )
      android.util.Log.d("RainAlert", "Notification shown")
    } catch (e: SecurityException) {
      android.util.Log.e("RainAlert", "Failed to show notification", e)
    }
  }
}

